### DISCUSSION TITLE
It can't get time tag data in loop.

### DISCUSSION THREAD

**Discussion by coderJoon** [PicoQuant/snAPI#15] [Q&A]
When I ran the demo code before deploying snAPI, it worked well, although I couldn't get data for about 10 out of 5000 loops.
However, after snAPI was deployed, I changed the code and ran it. As a result, **only 6 out of 10 loops were measured, and only 2 of them had time tag data.**  And, Compared to the demo code, **the photon number is also lacking a lot**. (NumRec in sn.measDescription) 

Here is my code. **Can you tell me where is wrong in my code?** My equipment is MultiHarp 150, acqtime = 80ms , meascrtl= C1 start ctc stop , measMode = T2
![image](https://github.com/PicoQuant/snAPI/assets/87929279/2b492d6b-457b-4834-96e6-83c02f980e1a)
![image](https://github.com/PicoQuant/snAPI/assets/87929279/ed1d243f-2754-407a-b856-85a4fa944bb1)

**Comment by tpoint75** [comment]
Hi coderJoon,

that sounds strange. Please provide the corresponding log file from c:\data\picoquant\logs. Are you using the latest release?

Hint: 
To print something to the stdout / log use sn.logprint();
sn.measDescription['NumRecs'] gives the number for raw data records including overflow records and so on. The exact number of photons you get with len(times).

Best regards

**Reply by coderJoon** [reply]
Thanks for the information about the number of photons.

I'm using version 1.0.2  --> Should i update it  to 1.0.3 ?

The log file is quite long, but I will attach it.

[240201.log](https://github.com/PicoQuant/snAPI/files/14271288/240201.log)

**Comment by tpoint75** [comment]
I understand now, your count rates are to small that the hardware collects not enough data for sending to the API.
I will talk to the HW developer abut it.
Anyway it takes some time to start and stop the measurement. Its maybe faster to let the measurement running and filter the data for the needed block:

    sn.unfold.startBlock(acqTime=0, size=1024*1024*1024, savePTU=False)
    
    for i in range(5000):
        sn.logPrint(f"Iteration: {i}")
        times = []
        channels = []
        finished = False
        while not finished:
            time.sleep(0.01)
            new_times, new_channels  = sn.unfold.getBlock()
            times.extend(new_times)
            channels.extend(new_channels)
            
            if len(times) > 0 and (times[-1] - times[0])/1e9 > 80:
                sn.logPrint((times[-1] - times[0])/1e9)
                break
        
        
        sn.getMeasDescription()
        sn.logPrint(f"Records: {sn.measDescription['NumRecs']} | {len(times)}")
        sn.logPrint(" channels |  times") 
        sn.logPrint("--------------------")
        
        for i in range(start,start+length):
            sn.logPrint(f"{channels[i]:9} | {times[i]:8}")
        
    sn.unfold.stopMeasure()

**Reply by coderJoon** [reply]
Thank you for your help.

To summarize to report to the prof. ,

is it true that the count rates are too low to load data through the API? and you will talk to your team and let me know later ?

**Reply by tpoint75** [reply]
Your code would working as it is, if the count rate would be higher and/or the acquisition time be longer.
Yes we will see, how to fix it, but this could take time.
Therefore I made a modified example, how to measure it with a different approach.
The advantage is that you can run faster 5000 iterations.

**Reply by coderJoon** [reply]
Thank you.

I have a question in the modified example.

what is "start" and "length" variable in last for loop?

**Reply by tpoint75** [reply]
Oh, I'm sorry start and length are only for the part of the records to be shown. You could take start=0 and length = 10 for the first ten records..

I talked with the HW developer and have a questions now:
Are you sure, you are using USB3 for the MultiHarp? Please Check the LED's on the front panel! 
Because it is not a problem of snAPI please start the original MultiHarp SW. Open menu about and press "Request Support" and "Open Support Page". There you can open a Ticket and Copy the Support Info Text. We need this Info.

Briefly describe your problem once again and tell that you are using snAPI. 

Regards, Torsten.

**Reply by coderJoon** [reply]
The LED panel emits clear green light.

Since I don't know all the equipment in the lab, I wrote down and submitted only the equipment I knew.

And I ran the example code that you gave me, it unable to receive data.

here is my code and log file (please look at the log starting at 13:09.52)
![image](https://github.com/PicoQuant/snAPI/assets/87929279/5dacae24-158e-4e21-b63d-4443dac60834)
[240219.log](https://github.com/PicoQuant/snAPI/files/14326294/240219.log)

**Reply by tpoint75** [reply]
The example was made to let the measurement running alle the time and only use the times and channel arrays for calculation, if needed. 
There for u have to set the acqTime = 0 (and not 80). The 80ms measurement time is now in the break condition inside the while loop..

**Reply by tpoint75** [reply]
Hello, do you have solved your problem? You have not answered on your support ticket!
